sls-versions/src/main/java/com/palantir/sls/versions/ReleaseVersionParser.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.sls.versions;
+
+import com.google.errorprone.annotations.Immutable;
+import com.palantir.sls.versions.MatchResult.Int3MatchResult;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+@Immutable
+enum ReleaseVersionParser implements Parser {
+    INSTANCE;
+
+    @Nullable
+    @Override
+    public MatchResult tryParse(String string) {
+        long state = Parsers.number(string, 0);
+        if (Parsers.failed(state)) {
+            return null;
+        }
+        int first = Parsers.getResult(state);
+
+        state = Parsers.literalDot(string, Parsers.getIndex(state));
+        if (Parsers.failed(state)) {
+            return null;
+        }
```

Comment:
This is just begging for some kind of monadic bind abstraction... but obviously that would require a lot of allocation in Java.

Reply:
_totally agree_